sql: fix internal executor when it encounters a retry error

yuzefovich opened the pull request:
This PR contains several commits that fix a long-standing bug in the internal executor that could make it double count some things (rows or metadata) if an internal retry error occurs. In particular, since at least 21.1 (when we introduced the "streaming" internal executor in #59330) if the IE encounters a retry error _after_ it has communicated some results to the client, it would proceed to retry the corresponding command as if the incomplete execution and the retry error never happened. In other words, it was possible for some rows to be double "counted" (either to be directly included multiple times into the result set or indirectly into the "rows affected" number). This PR fixes the problem by returning the retry error to the client in cases when some actual rows have already been communicated and by resetting the number of "rows affected" when "rewinding the stmt buffer" in order to retry the error transparently.

Fixes: #98558.

yuzefovich committed:
sql: minor cleanup around streamingCommandResult

This commit cleans up `streamingCommandResult` a bit by storing `CmdPos`
directly in it (which allows us to remove redundant `resWithPos`
wrapper) and removing a redundant argument in `closeCallback`. This is
effectively a purely mechanical change.

Also fix up some comments.

Release note: None

yuzefovich committed:
sql: use SetRowsAffected instead of IncrementRowsAffected

This commit changes the `RestrictedCommandResult` interface a bit to
make `SetRowsAffected` just set the number of rows affected, rather than
increment it. This method is supposed to be called only once, so this
change seems reasonable on its own, but it also has an additional
benefit for the follow-up commit. In particular, in the follow-up commit
we will preserve the ability to automatically retry statements of
non-ROWS stmt type by the internal executor.

Note that this commit on its own fixes the behavior for the "rows
affected" statements (the only known occurrence of the bug that is
generally fixed in the following commit).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/101477)
<!-- Reviewable:end -->

yuzefovich committed:
sql: transparently retry InternalExecutor.Exec{Ex} for ROWS stmts

This commit teaches the internal executor to avoid propagating the retry
errors to the client when executing `Exec{Ex}` methods. In those
methods, we only return the number of rows affected, thus, in order to
preserve the ability to transparently retry all statements (even if they
are of ROWS statement type), we simply need to be able to reset the rows
affected number on the `rowsIterator` if a result is being discarded
(which indicates that a retry error has been encountered and the rewind
mechanism is at play).

This structure relies on the assumption that we push at most one command
into the StmtBuf that results in "rows affected" which is true at the
moment of writing, and I don't foresee that changing, at least for now.
(This assumption would be incorrect only if we were to allow executing
multiple statements as part of a single method invocation on the
internal executor, but I don't think that's going to happen - we only
might do so in a few tests.)

Release note: None

yuzefovich commented:
@rafiss @knz I think two of you have the most context around this code, so I'm asking for review from both of you; @cucaroach I think you might have looked into this stuff recently, so I'd welcome your review too. Extra scrutiny is much appreciated given the trickiness and ubiquitousness of the internal executor.

yuzefovich commented:
Oh, I think @ZhouXing19 also has exposure - welcome your review too!

knz commented:
Amazing work. I'll review this shortly.

knz reviewed (commented):
For the 2nd commit, could I interest you in a test (something like the one from the 3rd commit) which checks that the rowsaffected are at the value expected even when a retry has occurred?

I haven't fully reviewed the changes to sql/internal.go yet because I find them intimidating :)

> Reviewed 4 of 4 files at r1, 7 of 7 files at r2, 3 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-:-NSysmDi9g4MZLriyfLU:b-8h11hg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @rafiss, @yuzefovich, and @ZhouXing19)

___
*[`-- commits` line 60 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NSyt6ew6Ol3DMww12Rs):*
For columns I'm not sure -- a retry could potentially bring a txn over a schema change that alters the number of columns?
___
*[`-- commits` line 76 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NSytkeX9k_QhCSNaCta):*
What happens if Exec shares a kv.Txn with the caller? In that case, if there's a retry error, it's not correct to just retry inside internalexecutor; because the caller may have performed some other SQL using the same kv.Txn prior to calling Exec, and all of that needs to be retried too.

I think we are only allowed to retry things inside internalexecutor if the caller specified a nil kv.Txn to start with. Where is this checked?


<!-- Sent from Reviewable.io -->

miretskiy commented:
should this be backported to 22.2?

cucaroach commented on pkg/sql/internal_test.go:
```diff
@@ -714,6 +714,23 @@ func TestInternalExecutorRetryAfterRows(t *testing.T) {
+	// Now verify that ExecEx correctly and transparently to us retries the
+	// stmt.
+	numRows, err := ie.ExecEx(
+		ctx, "read rows", nil, /* txn */
```
Should we have a test where txn isn't nil and has already been used to do something and we don't retry in that case?

cucaroach commented:
Looks good to me, nice work figuring this out!   I'll defer approval to others more familiar with this area.

yuzefovich committed:
sql: fix internal executor when it encounters a retry error

This commit fixes a bug with the internal executor when it encounters an
internal retry. Previously, the implementation of the "rewind
capability" was such that we always assumed that we can rewind the
StmtBuf for the IE at any point; however, that is not generally true. In
particular, if we communicated anything to the client of the IE's
connExecutor (`rowsIterator` in this context), then we cannot rewind the
current command we're evaluating. In theory, we could have pushed some
rows through the iterator only to encounter the internal retry later which
would then lead to re-executing the command from the start (in other
words, the rows that we've pushed before the retry would be
double-pushed); in practice, however, we haven't actually seen this (at
least yet). What we have seen is the number of rows affected being
double counted. This particular reproduction was already fixed in the
previous commit, but this commit fixes the problem more generally.

This commit makes it so that for every `streamingCommandResult` we are
tracking whether it has already communicated something to the client so
that the result can no longer be rewound, and then we use that tracking
mechanism to correctly implement the rewind capability. We have three
possible types of data that we communicate:
- rows
- number of rows affected
- column schema.

If the retry happens after some rows have been communicated, we're out
of luck - there is no way we can retry the stmt internally, so from now
on we will return a retry error to the client.

If the retry happens after "rows affected", then given the adjustment in
the previous commit we can proceed transparently.

In order to avoid propagating the retry error up when it occurs after
having received the column schema but before pushing out any rows, this
commit adjusts the behavior to always keep the latest column schema,
thus, we can still proceed transparently in this case.

This bug has been present since at least 21.1 when
`streamingCommandResult` was introduced. However, since we now might
return a retry error in some cases, this could lead to test failures or
flakes, or even to errors in some internal CRDB operations that execute
statements of ROWS type (if there is no appropriate retry logic), so
I intend to only backport this to 23.1. There is also no release note
since the only failure we've seen is about double counted "rows
affected" number, the likelihood of which has significantly increased
due to the jobs system refactor (i.e. mostly 23.1 is affected AFAIK).

Additionally, this commit makes it so that we correctly block the
`execInternal` call until the first actual, non-metadata result is seen
(this behavior is needed to correctly synchronize access to the txn
before the stmt is given to the execution engine).

Release note: None

yuzefovich reviewed (commented):
> For the 2nd commit, could I interest you in a test (something like the one from the 3rd commit) which checks that the rowsaffected are at the value expected even when a retry has occurred?

Done.

> should this be backported to 22.2?

I'm not sure yet, this stuff is scary to change. I think the second commit can and should be backported, but more general fix in the third and fourth commits are TBD.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-:-NT0D3Cr68LDG_S668fK:br18bbh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @knz, @rafiss, and @ZhouXing19)

___
*[`-- commits` line 60 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NSyt6ew6Ol3DMww12Rs):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

For columns I'm not sure -- a retry could potentially bring a txn over a schema change that alters the number of columns?
</blockquote></details>

Hm, this is a good question. It sounds plausible that this scenario can be problematic, so I tried to reproduce it in a test with the following:

<details>
<summary>code diff</summary>

```
diff --cc pkg/sql/internal_test.go
index 46ec7d059f3,79ef7fa5a95..00000000000
--- a/pkg/sql/internal_test.go
+++ b/pkg/sql/internal_test.go
@@@ -693,8 -697,26 +697,28 @@@ func TestInternalExecutorEncountersRetr
        defer leaktest.AfterTest(t)()
        defer log.Scope(t).Close(t)
  
+       const rowsStmt = `SELECT * FROM test.t`
+       var blockRowPush atomic.Bool
 -      blockPush := make(chan struct{})
++      rowPushed := make(chan struct{})
++      unblockPush := make(chan struct{})
+ 
        ctx := context.Background()
        params, _ := tests.CreateTestServerParams()
+       params.Knobs.SQLExecutor = &sql.ExecutorTestingKnobs{
+               DistSQLReceiverPushCallbackFactory: func(query string) func(rowenc.EncDatumRow, coldata.Batch, *execinfrapb.ProducerMetadata) {
+                       if query != rowsStmt || !blockRowPush.Load() {
+                               return nil
+                       }
+                       var blockedPush bool
+                       return func(row rowenc.EncDatumRow, _ coldata.Batch, _ *execinfrapb.ProducerMetadata) {
+                               if row != nil && !blockedPush {
 -                                      <-blockPush
++                                      rowPushed <- struct{}{}
++                                      <-unblockPush
+                                       blockedPush = true
+                               }
+                       }
+               },
+       }
        s, db, _ := serverutils.StartServer(t, params)
        defer s.Stopper().Stop(ctx)
  
@@@ -739,18 -760,15 +762,40 @@@
                }
        })
  
 +      // This test case verifies that ExecEx of a stmt of Rows type correctly and
 +      // transparently to us retries the stmt.
 +      t.Run("ExecEx retries", func(t *testing.T) {
-               const stmt = "SELECT * FROM test.t"
-               numRows, err := ie.ExecEx(ctx, "read rows", nil /* txn */, ieo, stmt)
++              numRows, err := ie.ExecEx(ctx, "read rows", nil /* txn */, ieo, rowsStmt)
 +              if err != nil {
 +                      t.Fatal(err)
 +              }
 +              if numRows != 1 {
 +                      t.Fatalf("expected 1 rowsAffected, got %d", numRows)
 +              }
 +      })
++
+       t.Run("schema change in-between retries", func(t *testing.T) {
+               blockRowPush.Store(true)
+               defer blockRowPush.Store(false)
+ 
+               errCh := make(chan error)
+               go func() {
 -
++                      _, err := ie.ExecEx(ctx, "read rows", nil /* txn */, ieo, rowsStmt)
++                      errCh <- err
+               }()
++
++              <-rowPushed
++
++              dbConn := serverutils.OpenDBConn(
++                      t, s.ServingSQLAddr(), "" /* useDatabase */, false /* insecure */, s.Stopper(),
++              )
++              defer dbConn.Close()
++              _, err := dbConn.Exec(`ALTER TABLE test.t ADD COLUMN c2 INT NULL`)
++
++              unblockPush <- struct{}{}
++              err = <-errCh
++              t.Fatal(err)
+       })
  }
  
  // TODO(andrei): Test that descriptor leases are released by the
```

</details>

This test starts a query via `ExecEx` (i.e. with "rows affected" execution mode) but blocks the execution before it pushes the first row. Then in a separate goroutine we try to modify the table by adding a column. This schema change though seems to never complete - it hits retry errors internally (not injected by the test) when attempting to commit. Even if I unblock the query after some timeout, things still seem stuck.

Perhaps we don't have to think about this case too much though? My reasoning is as follows: we are evaluating the stmt of ROWS type (which is the only case when this column schema matters since it's provided by the `rowsIterator`). Then, if a retry occurs before any rows are pushed, then the query can be safely retried as long as the `rowsIterator` keeps the last column schema it has seen; if a retry occurs after some rows have been pushed, then the query results in a retry error anyway.

Currently, the first column schema is kept by the `rowsIterator`, so I adjusted the third commit to fix that (I hope it doesn't break anything :crossed_fingers:). I'm having hard time coming up with a regression test for it though.
___
*[`-- commits` line 76 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NSytkeX9k_QhCSNaCta):*
> I think we are only allowed to retry things inside internalexecutor if the caller specified a nil kv.Txn to start with. Where is this checked?

Yes, this is correct. We can only retry in implicit txns. The checking is done in `connExecutor.makeErrEvent` where we call `getRewindTxnCapability` only for implicit txns. What happens with an explicit txn is the following:
- we have `ExecStmt` which encounters a retry
- the error is stored in `icc.results[0]` but is not yet communicated back to the client
- the connExecutor's state machine rolls back the stmt
- then it processes the `Sync` command, where in `syncCallback` we push out `[2]*streamingCommandResults` with the zeroth having the error
- now the client reads the results, finds the error and returns it back to the caller of IE.
___
*[`pkg/sql/internal.go` line 1112 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/101477#gh-1167104226:-NT0Fev883B__cj8_Ano:b-c71iyx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f183f6d20b6f9b5dcaef788987cca36dfae7ff7f/pkg/sql/internal.go#L1112)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

Why is it the case that only one command in the stmtbuf results in row affected?
</blockquote></details>

We are only allowed to execute a single stmt via one invocation of the internal executor, and we either use `ExecStmt, Sync` commands (when there are no query arguments) or `PrepareStmt, BindStmt, ExecPortal, Sync` (when we have query arguments). In both of these sequences only `ExecStmt` or `ExecPortal` actually run via the execution engine which is when we populate "rows affected" info.
___
*[`pkg/sql/internal_test.go` line 721 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/101477#gh-1167105477:-NT0l1PLEu-QhDfvDicg:b4s70jm) ([raw file](https://github.com/cockroachdb/cockroach/blob/f183f6d20b6f9b5dcaef788987cca36dfae7ff7f/pkg/sql/internal_test.go#L721)):*
<details><summary><i>Previously, cucaroach (Tommy Reilly) wrote…</i></summary><blockquote>

Should we have a test where txn isn't nil and has already been used to do something and we don't retry in that case?
</blockquote></details>

Good point, I added a test with non-nil txn. We don't actually retry ever with an explicit txn, regardless of whether txn did something or not.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
commits 1,2,4 i understood better and lgtm with some nits! thanks for the fix. commit 3 i think seems good too, but i'm less familiar. i can review it closer after i'm back from PTO, but don't let me block anything if that's not ideal

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-:-NT2_Lch2c1sXyG7qLsg:baejzop)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @knz, @yuzefovich, and @ZhouXing19)

___
*[`pkg/sql/sessiondata/internal.go` line 49 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NT2YLvJERLCZ8ucGvwm:-NT2YLvJERLCZ8ucGvwn:b3ymprc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a8e51f2b966d86636237dbbe317e329c8def6b5/pkg/sql/sessiondata/internal.go#L49)):*
> ```Go
> 	// does **not** propagate further to "nested" executors that are spawned up
> 	// by the "top" executor.
> 	InjectRetryErrorsEnabled bool
> ```

nice to see this being useful internally too!
___
*[`pkg/sql/sessiondatapb/local_only_session_data.proto` line 207 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NT2YSuBBjcWhCne0zaO:-NT2YSuCCy042iHJi__c:blbj5o0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a8e51f2b966d86636237dbbe317e329c8def6b5/pkg/sql/sessiondatapb/local_only_session_data.proto#L207)):*
> ```protobuf
>   // developers to test their app's retry logic.
>   //
>   // Note that this session variable is **not** propagated to the internal
> ```

it should be the case that it does get propagated if you use an internal executor in the context of a user's session. but if you use it as part of a background job, then there is no session data to propagate from. or is there some other bug that is preventing it from being inherited even when it should?

semi-related, i've been working on https://github.com/cockroachdb/cockroach/pull/101486 to improve how background jobs' session data is created
___
*[`pkg/sql/internal_test.go` line 717 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NT2Yrcz1bQydJ1DPHne:-NT2Yrcz1bQydJ1DPHnf:b4p0m8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a8e51f2b966d86636237dbbe317e329c8def6b5/pkg/sql/internal_test.go#L717)):*
> ```Go
> 			ctx, "pause schedule", nil, /* txn */
> 			sessiondata.InternalExecutorOverride{
> 				User:                     username.MakeSQLUsernameFromPreNormalizedString(username.RootUser),
> ```

nit: use username.RootUserName() instead


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
Excellent work, I learned a lot :smile:  !
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod nit comments, but I'd like to defer others to finalize the review

> Reviewed 4 of 4 files at r1, 1 of 10 files at r5, 7 of 10 files at r6, 3 of 3 files at r7, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-:-NTAJe4y-i_s7dYjeBxs:bjnlk3g)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach, @knz, and @yuzefovich)

___
*[`-- commits` line 60 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NSyt6ew6Ol3DMww12Rs):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, this is a good question. It sounds plausible that this scenario can be problematic, so I tried to reproduce it in a test with the following:

<details>
<summary>code diff</summary>

```
diff --cc pkg/sql/internal_test.go
index 46ec7d059f3,79ef7fa5a95..00000000000
--- a/pkg/sql/internal_test.go
+++ b/pkg/sql/internal_test.go
@@@ -693,8 -697,26 +697,28 @@@ func TestInternalExecutorEncountersRetr
        defer leaktest.AfterTest(t)()
        defer log.Scope(t).Close(t)
  
+       const rowsStmt = `SELECT * FROM test.t`
+       var blockRowPush atomic.Bool
 -      blockPush := make(chan struct{})
++      rowPushed := make(chan struct{})
++      unblockPush := make(chan struct{})
+ 
        ctx := context.Background()
        params, _ := tests.CreateTestServerParams()
+       params.Knobs.SQLExecutor = &sql.ExecutorTestingKnobs{
+               DistSQLReceiverPushCallbackFactory: func(query string) func(rowenc.EncDatumRow, coldata.Batch, *execinfrapb.ProducerMetadata) {
+                       if query != rowsStmt || !blockRowPush.Load() {
+                               return nil
+                       }
+                       var blockedPush bool
+                       return func(row rowenc.EncDatumRow, _ coldata.Batch, _ *execinfrapb.ProducerMetadata) {
+                               if row != nil && !blockedPush {
 -                                      <-blockPush
++                                      rowPushed <- struct{}{}
++                                      <-unblockPush
+                                       blockedPush = true
+                               }
+                       }
+               },
+       }
        s, db, _ := serverutils.StartServer(t, params)
        defer s.Stopper().Stop(ctx)
  
@@@ -739,18 -760,15 +762,40 @@@
                }
        })
  
 +      // This test case verifies that ExecEx of a stmt of Rows type correctly and
 +      // transparently to us retries the stmt.
 +      t.Run("ExecEx retries", func(t *testing.T) {
-               const stmt = "SELECT * FROM test.t"
-               numRows, err := ie.ExecEx(ctx, "read rows", nil /* txn */, ieo, stmt)
++              numRows, err := ie.ExecEx(ctx, "read rows", nil /* txn */, ieo, rowsStmt)
 +              if err != nil {
 +                      t.Fatal(err)
 +              }
 +              if numRows != 1 {
 +                      t.Fatalf("expected 1 rowsAffected, got %d", numRows)
 +              }
 +      })
++
+       t.Run("schema change in-between retries", func(t *testing.T) {
+               blockRowPush.Store(true)
+               defer blockRowPush.Store(false)
+ 
+               errCh := make(chan error)
+               go func() {
 -
++                      _, err := ie.ExecEx(ctx, "read rows", nil /* txn */, ieo, rowsStmt)
++                      errCh <- err
+               }()
++
++              <-rowPushed
++
++              dbConn := serverutils.OpenDBConn(
++                      t, s.ServingSQLAddr(), "" /* useDatabase */, false /* insecure */, s.Stopper(),
++              )
++              defer dbConn.Close()
++              _, err := dbConn.Exec(`ALTER TABLE test.t ADD COLUMN c2 INT NULL`)
++
++              unblockPush <- struct{}{}
++              err = <-errCh
++              t.Fatal(err)
+       })
  }
  
  // TODO(andrei): Test that descriptor leases are released by the
```

</details>

This test starts a query via `ExecEx` (i.e. with "rows affected" execution mode) but blocks the execution before it pushes the first row. Then in a separate goroutine we try to modify the table by adding a column. This schema change though seems to never complete - it hits retry errors internally (not injected by the test) when attempting to commit. Even if I unblock the query after some timeout, things still seem stuck.

Perhaps we don't have to think about this case too much though? My reasoning is as follows: we are evaluating the stmt of ROWS type (which is the only case when this column schema matters since it's provided by the `rowsIterator`). Then, if a retry occurs before any rows are pushed, then the query can be safely retried as long as the `rowsIterator` keeps the last column schema it has seen; if a retry occurs after some rows have been pushed, then the query results in a retry error anyway.

Currently, the first column schema is kept by the `rowsIterator`, so I adjusted the third commit to fix that (I hope it doesn't break anything :crossed_fingers:). I'm having hard time coming up with a regression test for it though.
</blockquote></details>

I had the same concern, and I think the fix with looping updates for `r.resultCols` makes sense. However, the code diff above doesn't seem to simulate a schema change that happens between retrying -- IIUC here the schema change happens _before_ the results are pushed to `streamingCommandResult`, but shouldn't it happen after `res.SetError(retryErr)`? 
___
*[`pkg/sql/conn_io.go` line 1041 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NT9mB9k5AG15O_GBWkS:-NT9mB9lA-gioVVIP5-s:b-9ssbqa) ([raw file](https://github.com/cockroachdb/cockroach/blob/791e7c7432fa6ff9efd053499693fc8cc5b2116d/pkg/sql/conn_io.go#L1041)):*
> ```Go
> 	// Once we add this row to the writer, it can be immediately consumed by the
> 	// reader, so this result can no longer be rewound.
> 	r.cannotRewind = true
> ```

nit: if we're in `rowsAffectedIEExecutionMode`, we allow rewinding but still set `r.cannotRewind = true` here. Do you think we should have a gate here? In other words, maybe we can move the `icc.mode == rowsAffectedIEExecutionMode` from `icc.ClientPos()` to here. 


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks for the reviews! I'll definitely wait for Raphael's approval.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-:-NTH7Sce-FNdRrEwSyqC:bwo4jtl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @knz, @rafiss, and @ZhouXing19)

___
*[`-- commits` line 60 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NSyt6ew6Ol3DMww12Rs):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

I had the same concern, and I think the fix with looping updates for `r.resultCols` makes sense. However, the code diff above doesn't seem to simulate a schema change that happens between retrying -- IIUC here the schema change happens _before_ the results are pushed to `streamingCommandResult`, but shouldn't it happen after `res.SetError(retryErr)`? 
</blockquote></details>

Yeah, perhaps you're right. I tried doing that (we already have `AfterExecute` testing knob), and the test still seems deadlocked (i.e. the schema change seems to be stuck).
___
*[`pkg/sql/conn_io.go` line 1041 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NT9mB9k5AG15O_GBWkS:-NTH80Dz4gC8q5velZKX:b-7ugyik) ([raw file](https://github.com/cockroachdb/cockroach/blob/791e7c7432fa6ff9efd053499693fc8cc5b2116d/pkg/sql/conn_io.go#L1041)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

nit: if we're in `rowsAffectedIEExecutionMode`, we allow rewinding but still set `r.cannotRewind = true` here. Do you think we should have a gate here? In other words, maybe we can move the `icc.mode == rowsAffectedIEExecutionMode` from `icc.ClientPos()` to here. 
</blockquote></details>

Hm, I think I prefer the current approach more. In my mind, `internalClientComm` is the component that should be aware of `ieExecutionMode`, so methods on `icc` is where we should examine the mode. Pushing the mode into `streamingCommandResult` struct would work too, but that would mean spreading the logic of the `ieExecutionMode` over more surface area (we'd still need to keep `internalClientComm.mode` field for `createRes` method), making it harder to reason about.
___
*[`pkg/sql/sessiondatapb/local_only_session_data.proto` line 207 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NT2YSuBBjcWhCne0zaO:-NTH6RANEdDaTcJ24aQG:b-tyeg9v) ([raw file](https://github.com/cockroachdb/cockroach/blob/3a8e51f2b966d86636237dbbe317e329c8def6b5/pkg/sql/sessiondatapb/local_only_session_data.proto#L207)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

it should be the case that it does get propagated if you use an internal executor in the context of a user's session. but if you use it as part of a background job, then there is no session data to propagate from. or is there some other bug that is preventing it from being inherited even when it should?

semi-related, i've been working on https://github.com/cockroachdb/cockroach/pull/101486 to improve how background jobs' session data is created
</blockquote></details>

This session variable on its own works, but the change in the second commit makes it so that `InternalExecutorOverride` applies all the time, even if `InternalExecutorOverride.InjectRetryErrorsEnabled` wasn't specified. In other words, the second commit effectively disables this session variable for Internal Executors. I needed this behavior so that if we have an IE query that issues another IE query, then the latter doesn't get a retry injection. This seems like a reasonable behavior. WDYT?


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
I am still ok with most of the changes here.
Sadly I find it very difficult to review the changes to conn_io.go and internal.go because I do not understand them. I do not have a mental model of what the code was doing prior to your change; nor do I have a mental model of what your changes do.

For me to do this review successfully, I'd need a deeper explanation of how the code worked before, what is being changed here and why you're expecting to work.

> Reviewed 1 of 10 files at r5, 7 of 10 files at r6, 3 of 3 files at r9, 1 of 3 files at r10, 1 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-:-NTI7gMD6nSQI9hSj70X:b-4hahfh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @rafiss, @yuzefovich, and @ZhouXing19)

___
*[`-- commits` line 60 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NSyt6ew6Ol3DMww12Rs):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Yeah, perhaps you're right. I tried doing that (we already have `AfterExecute` testing knob), and the test still seems deadlocked (i.e. the schema change seems to be stuck).
</blockquote></details>

Can we get some help from Faizan or someone else in the schema team to get a working test.
___
*[`-- commits` line 76 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NSytkeX9k_QhCSNaCta):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

> I think we are only allowed to retry things inside internalexecutor if the caller specified a nil kv.Txn to start with. Where is this checked?

Yes, this is correct. We can only retry in implicit txns. The checking is done in `connExecutor.makeErrEvent` where we call `getRewindTxnCapability` only for implicit txns. What happens with an explicit txn is the following:
- we have `ExecStmt` which encounters a retry
- the error is stored in `icc.results[0]` but is not yet communicated back to the client
- the connExecutor's state machine rolls back the stmt
- then it processes the `Sync` command, where in `syncCallback` we push out `[2]*streamingCommandResults` with the zeroth having the error
- now the client reads the results, finds the error and returns it back to the caller of IE.
</blockquote></details>

thanks


<!-- Sent from Reviewable.io -->

yuzefovich committed:
sql: document how the machinery of the internal executor fits together

Release note: None

yuzefovich reviewed (commented):
That's a fair point. I can't say that I fully grokked how all this machinery works when I implemented the "streaming" internal executor a couple years ago. Now that I spent significant amount of time debugging this code, I have a decent mental model, so I just added another commit (now second) which attempts to outline how all moving pieces fit together, and then with each consequent commit I tried to update the corresponding comment accordingly.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-:-NTLs5s_A9ksVsIwhbVF:b2ovybe)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @knz, @rafiss, and @ZhouXing19)

___
*[`-- commits` line 60 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NSyt6ew6Ol3DMww12Rs):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Can we get some help from Faizan or someone else in the schema team to get a working test.
</blockquote></details>

Do you think this is a blocking question? I'm not saying we shouldn't introduce a test like this, but it seems like an edge case that we can leave a TODO for.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
This new doc is very good. Thank you.

> Reviewed 1 of 1 files at r13, 1 of 10 files at r14, 2 of 3 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-:-NTPEnxK9h0GmVVWgpGI:bu3to7h)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @rafiss, @yuzefovich, and @ZhouXing19)

___
*[`-- commits` line 60 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NSyt6ew6Ol3DMww12Rs):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Do you think this is a blocking question? I'm not saying we shouldn't introduce a test like this, but it seems like an edge case that we can leave a TODO for.
</blockquote></details>

No it is not blocking this PR, but we still need this test to happen.
___
*[`pkg/sql/internal.go` line 818 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NTP-aPS6RXepADkQ10Z:-NTP-aPS6RXepADkQ10_:b-7en63j) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fcfcf0a258960093ac858f5a82f02c0ab325a25/pkg/sql/internal.go#L818)):*
> ```Go
> // It's also worth noting that execInternal doesn't return until the
> // connExecutor reaches the execution engine (i.e. until after the query
> // planning has been performed). This is needed in order to avoid concurrent
> ```

I don't understand why this fully prevents concurrent access to the kv.Txn. If more rows are returned to the caller via the iterator, aren't those rows the product of a currently-used kv.Txn? What if there's another call to IE "under" the iterator?
___
*[`pkg/sql/internal.go` line 820 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NTP-XZX0nsD29OdKHTn:-NTP-XZX0nsD29OdKHTo:b-46vrj9) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fcfcf0a258960093ac858f5a82f02c0ab325a25/pkg/sql/internal.go#L820)):*
> ```Go
> // planning has been performed). This is needed in order to avoid concurrent
> // access to the txn.
> // TODO(yuzefovich): currently, this statement is not entirely true if the retry
> ```

I'm curious - what is not true about it?
___
*[`pkg/sql/internal.go` line 833 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NTP0QMo3K4kQPwmrwYF:-NTP0QMpE67hL1ZOzUaM:b-rvgrwh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fcfcf0a258960093ac858f5a82f02c0ab325a25/pkg/sql/internal.go#L833)):*
> ```Go
> // internalClientComm implements the ClientLock interface in such a fashion as
> // if any command can be transparently retried.
> // TODO(yuzefovich): fix this.
> ```

Maybe worth filing issues and linking them from the comment.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks for the reviews everyone!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-:-NTPLZR60kgX5OpWsYOT:bjbxjiq)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @knz, @rafiss, and @ZhouXing19)

___
*[`-- commits` line 60 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NSyt6ew6Ol3DMww12Rs):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

No it is not blocking this PR, but we still need this test to happen.
</blockquote></details>

Left a TODO for now.
___
*[`pkg/sql/internal.go` line 818 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NTP-aPS6RXepADkQ10Z:-NTPJYANES8CTgpq2KUn:b-5xwnog) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fcfcf0a258960093ac858f5a82f02c0ab325a25/pkg/sql/internal.go#L818)):*
It prevents the concurrent access between two goroutines at play - the `rowsIterator` and the "internal" `connExecutor` goroutines. If the "synchronous" `ieResultChannel` is used, only one goroutine will be active at any point in time. When more rows are returned to the caller, the caller was blocked on `ieResultChannel` while the rows were being produced.

> What if there's another call to IE "under" the iterator?

This question is spot on - we actually put this synchronization in-place due to a bug with "nested" IE #62415. I expanded the comment.
___
*[`pkg/sql/internal.go` line 820 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NTP-XZX0nsD29OdKHTn:-NTPKjsn5uXxqfKx-7z-:b-7lnlri) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fcfcf0a258960093ac858f5a82f02c0ab325a25/pkg/sql/internal.go#L820)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

I'm curious - what is not true about it?
</blockquote></details>

The problem was that if the retry occurs after the column schema was sent but before any rows were communicated, this call wouldn't block until the control flow is in the execution engine (when reevaluating the stmt after the retry). In other words, if we had a sequence like
- send columns
- encounter a retry
- send columns (1)
- go into execution engine (2)

the blocking was done only up until (1), but it should be up until (2). This is fixed in the fourth commit by looping until we see non-column data.
___
*[`pkg/sql/internal.go` line 833 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/101477#-NTP0QMo3K4kQPwmrwYF:-NTPLMHR8zv4w1cXJ0Wu:bwtfglk) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fcfcf0a258960093ac858f5a82f02c0ab325a25/pkg/sql/internal.go#L833)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Maybe worth filing issues and linking them from the comment.
</blockquote></details>

This is fixed in the fourth commit to - this TODO is the general bug this PR is fixing.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Known flake #101519.

bors r+
